Add GKE openapi2 Support; CRD example

salrashid123 opened the pull request:
Fixes https://github.com/GoogleCloudPlatform/deploymentmanager-samples/issues/48

* Adds support for openapi2 GKE spec 
* Maps `name` and `namespace` directly from `metadata` values per [specs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata)
* Adds basic example for `customresourcedefinitions`

salrashid123 commented:
hmm..while the `CRD` definitions are created, i havne't figured out how to actually create a crd object itself.... 

It seems DM acquires  whats at `/openapi/v2`  on create and k8s endpoint doesn't list out the new CRDs there (i.,e even if i create a crd definition, its not listed there)

--

for example even if i create a CRD at
`/apis/stable.example.com/v1/namespaces/default/crontabs/{name}`

i can't create an object by specifying a yaml like
- crontab.jinja
```yaml
{% set CLUSTER_TYPE = env['project'] + '/' + properties['clusterType'] %}
{% set CRONTAB_COLLECTION =  '/apis/stable.example.com/v1/namespaces/default/crontabs/{name}' %}
{% set NAME_PREFIX = env['deployment'] + '-' + env['name'] %}

resources:
- name: {{ NAME_PREFIX }}-crd-crontab
  type: {{ CLUSTER_TYPE }}:{{ CRONTAB_COLLECTION }}
  properties:
    apiVersion: "stable.example.com/v1"
    kind: CronTab    
    metadata:
      namespace: default
      name: my-new-cron-object    
    spec:
      cronSpec: "* * * * /5"
      image: my-awesome-cron-image
```
which gives a not found
```
gcloud deployment-manager deployments create my-crd-crontab --template crontab.jinja --properties clusterType:${CLUSTER_NAME}-provider

ERROR: (gcloud.deployment-manager.deployments.create) Error in Operation [operation-1550553938231-5823878b18269-8e447d9b-7c4da7c1]: errors:
- code: COLLECTION_NOT_FOUND
  message: Collection '/apis/stable.example.com/v1/namespaces/{namespace}/crontabs/{name}'
    not found in discovery doc 'https://35.225.120.54/openapi/v2'
```

just to note, here's the endpoint kubectl uses which i have to figure out how DM can also use...

- `kubectl apply -f crd-object.yaml`

```
I0218 21:27:11.668957  139586 round_trippers.go:406] POST https://35.225.120.54/apis/stable.example.com/v1/namespaces/default/crontabs 201 Created in 113 milliseconds
I0218 21:27:11.669254  139586 request.go:874] Response Body: {"apiVersion":"stable.example.com/v1","kind":"CronTab","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"stable.example.com/v1\",\"kind\":\"CronTab\",\"metadata\":{\"annotations\":{},\"name\":\"my-new-cron-object\",\"namespace\":\"default\"},\"spec\":{\"cronSpec\":\"* * * * */5\",\"image\":\"my-awesome-cron-image\"}}\n"},"creationTimestamp":"2019-02-19T05:27:11Z","generation":1,"name":"my-new-cron-object","namespace":"default","resourceVersion":"7718","selfLink":"/apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object","uid":"0217a930-3407-11e9-bb35-42010a800037"},"spec":{"cronSpec":"* * * * */5","image":"my-awesome-cron-image"}}

crontab.stable.example.com "my-new-cron-object" created
```

Capstan commented:
Can this update be added in now? 1.14 is now available on GKE and I wager 1.13 will not be the default long, and all these samples will stop working.

I also have the same issue, where I'd like to get the OpenAPI v2 data for BackendConfig, but it's not served at the /openapi/v2 path AFAICT. And I'm not sure how to create a type provider from the results of getting the CRD resource itself, especially since it's OpenAPI v3.

Capstan commented:
Apparently, it's [an alpha feature `CustomResourcePublishOpenAPI` in 1.14](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2), so that's why I don't have CRDs.

salrashid123 commented:
i was able to get CRD defined and instantiated fine back in june while specifying `initialClusterVersion: 1.14.1-gke.5`..however, i ran into another issue with DM:  there seems to be some sort of bizzarre character limit while parsing the swagger specs only _after_ the CRD definitions are setup.   @Capstan   if you want, the internal# is b/133861742

This is what i did

```
NAME="dm-1"
CLUSTER_NAME="dm-gke-cluster-1"
ZONE="us-east1-d"


# 1. define GKE cluster
gcloud deployment-manager deployments create ${NAME} --template cluster.jinja --properties clusterName:${CLUSTER_NAME},zone:${ZONE}

gcloud beta deployment-manager type-providers describe ${CLUSTER_NAME}-provider --format='value(descriptorUrl)'

gcloud container clusters get-credentials $CLUSTER_NAME --zone $ZONE


# 2. try to get the master and token so we can 'view' the swagger specs before and after we setup the crd

## Before
MASTER_IP=`gcloud container clusters describe $CLUSTER_NAME --zone $ZONE --format='value(endpoint)'`
echo $MASTER_IP

curl -sk -H "Authorization: Bearer `gcloud auth print-access-token`" https://$MASTER_IP/openapi/v2 > swagger1.json



## 3. Now define the CRD
gcloud deployment-manager deployments create my-crd --template crd.jinja --properties clusterType:${CLUSTER_NAME}-provider

gcloud beta deployment-manager type-providers describe ${CLUSTER_NAME}-provider

### you should see the crd's youv'e defined here:
curl -sk -H "Authorization: Bearer `gcloud auth print-access-token`" https://$MASTER_IP/openapi/v2 > swagger2.json


>>>> 4.  At this moment, DM does not *know* about the new resource paths that just got crated so we need to ask it to 'refresh' its view of valid paths:

gcloud beta deployment-manager type-providers update ${CLUSTER_NAME}-provider --descriptor-url=https://$MASTER_IP/openapi/v2 --api-options-file=api-options.yaml

### 5. now that DM knows about the new paths, create an instance
gcloud deployment-manager deployments create my-crd-instance --template crd-instance.jinja --properties clusterType:${CLUSTER_NAME}-provider

```

Step 4. will fail though....its becasue the new CRD definition (`swagger2.json`) has text that exceeds a certain number described inthe bug.   I did confirm its the character limit because i did download swagger2.json, remove the extra character counts, then uploaded it to 'fake' --descriptor-url (which isn't GKE's master server)...then ran step 5.

All that is to say is we need to fix the character parsing...and even after that, it'd be nice if DM could 'auto refresh' its view of the rsource server so i don't have to do step 4 anyway


https://gist.github.com/salrashid123/65649168a6ec7ae4f8adaffd83cfe737

salrashid123 commented:
i just reverified it end-to-end here by forcing a swagger spec DM's parser likes.


This is what i did..you can find the source here:
  https://gist.github.com/salrashid123/65649168a6ec7ae4f8adaffd83cfe737

the only reason i edited the swagger specs below was just to see it actually work (slowly since DM does repeated lookups of the same specs seemingly unnecessarily).  the parsing length thing needs to be fixed in DM ofcourse


---

- export variables
```
NAME="dm-1"
CLUSTER_NAME="dm-gke-cluster-1"
ZONE="us-central1-a"
```

#### Create GKE Cluster

```
$ gcloud deployment-manager deployments create ${NAME} --template cluster.jinja --properties clusterName:${CLUSTER_NAME},zone:${ZONE}
                                               
    Create operation operation-1572649526816-596510066ee39-ad60d1e9-5fef6c40 completed successfully.
    NAME                       TYPE                                   STATE      ERRORS  INTENT
    dm-gke-cluster-1           container.v1.cluster                   COMPLETED  []
    dm-gke-cluster-1-provider  deploymentmanager.v2beta.typeProvider  COMPLETED  []
    OUTPUTS      VALUE
    clusterType  dm-gke-cluster-1-provider
```

### Get DM Type Provider URL  and MASTER_IP

```
$ gcloud beta deployment-manager type-providers describe ${CLUSTER_NAME}-provider --format='value(descriptorUrl)'
    https://34.67.124.249/openapi/v2


$ gcloud container clusters get-credentials $CLUSTER_NAME --zone $ZONE

$ MASTER_IP=`gcloud container clusters describe $CLUSTER_NAME --zone $ZONE --format='value(endpoint)'`

$ echo $MASTER_IP
    34.67.124.249

```

#### Create CRD

```
$ gcloud deployment-manager deployments create my-crd --template crd.jinja --properties clusterType:${CLUSTER_NAME}-provider
                                               
    Create operation operation-1572650044683-596511f44f6d8-cf1c9c43-410e1bd5 completed successfully.
    NAME                  TYPE                                                                                                               STATE      ERRORS  INTENT
    my-crd-crd-jinja-crd  mineral-minutia-820/dm-gke-cluster-1-provider:/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/{name}  COMPLETED  []


$  kubectl get crd
NAME                                    CREATED AT
crontabs.stable.example.com             2019-11-01T23:18:47Z



$ gcloud beta deployment-manager type-providers describe ${CLUSTER_NAME}-provider
```



#### Modify and override Swagger file

- Get the swagger specs:

```
$ curl -sk -H "Authorization: Bearer `gcloud auth print-access-token`" https://$MASTER_IP/openapi/v2 > swagger2.json
```

- remove the extra text that causes DM parsing error
 
```
$ sed -i 's/The continue option should be set when retrieving more results from the server. Since this value is server defined, clients may only use the continue value from a previous query result with identical query parameters (except for the value of continue) and the server may reject a continue value it does not recognize. If the specified continue value is no longer valid whether due to expiration (generally five to fifteen minutes) or a configuration change on the server, the server will respond with a 410 ResourceExpired error together with a continue token. If the client needs a consistent list, it must restart their list without the continue field. Otherwise, the client may send another list request with the token received with the 410 error, the server will respond with a list starting from the next key, but from the latest snapshot, which is inconsistent from the previous list results - objects that are created, modified, or deleted after the first list request will be included in the response, as long as their keys are after the \\"next key\\".\\n\\nThis field is not supported when watch is true. Clients may start a watch from the last resourceVersion value returned by the server and not miss any modifications./ /g' swagger2.json

```

- edit swagger2.json and add Host back to the MASTER_IP

```
 {"swagger":"2.0","host": "34.67.124.249","schemes":["https"],
```


- upload swagger2.json somewhere thats accessible

(for me, i used a public GCS bucket here)

```
$ gsutil cp swagger2.json gs://gcs.esodemoapp2.com/swagger2.json
```


- verify "its live"

```
$ curl -s http://gcs.esodemoapp2.com/swagger2.json | jq '.' | head
{
  "swagger": "2.0",
  "host": "34.67.124.249",
  "schemes": [
    "https"
  ],
```

#### Update DM to use the new specs

```
$ gcloud beta deployment-manager type-providers update ${CLUSTER_NAME}-provider \
    --descriptor-url=http://gcs.esodemoapp2.com/swagger2.json --api-options-file=api-options.yaml
                                           
    Updated type_provider [dm-gke-cluster-1-provider].
```

```
$ gcloud deployment-manager deployments list
    NAME                      LAST_OPERATION_TYPE  STATUS  DESCRIPTION  MANIFEST                ERRORS
    dm-1                      insert               DONE                 manifest-1572649527023  []
    my-crd                    insert               DONE                 manifest-1572650044724  []

$ kubectl get crd
    NAME                                    CREATED AT
    crontabs.stable.example.com             2019-11-01T23:18:47Z

$ kubectl get crontab
    No resources found.
```

#### Create the CRD instance

```
$ gcloud deployment-manager deployments create my-crd-instance --template crd-instance.jinja --properties clusterType:${CLUSTER_NAME}-provider
                                         
    Create operation operation-1572656117541-59652893d691b-d55e4860-fbbca360 completed successfully.
    NAME                                             TYPE                                                                                                              STATE      ERRORS  INTENT
    my-crd-instance-crd-instance-jinja-crd-instance  mineral-minutia-820/dm-gke-cluster-1-provider:/apis/stable.example.com/v1/namespaces/{namespace}/crontabs/{name}  COMPLETED  []
```


```
$ gcloud deployment-manager deployments list
    dm-1                      insert               DONE                 manifest-1572649527023  []
    my-crd                    insert               DONE                 manifest-1572650044724  []
    my-crd-instance           insert               DONE                 manifest-1572656117578  []

$ kubectl get crd
    NAME                                    CREATED AT
    crontabs.stable.example.com             2019-11-01T23:18:47Z

$ kubectl get crontab
    NAME                 AGE
    my-new-cron-object   2m30s <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```

#### Delete 

```
$ gcloud deployment-manager deployments delete my-crd-instance -q                                           
    Delete operation operation-1572656420441-596529b4b4c86-67397867-3b66e3cb completed successfully.

$ kubectl get crontab
    No resources found.

 ```

Capstan commented:
I was able to create CRDs via the following `.jinja` file using an alpha 1.14 cluster with https://gist.github.com/Capstan/797ebcc97a2d3089f563e33cbab15f5d. I haven't diagnosed the differences between your jist and mine yet.

salrashid123 commented:
@Capstan It wasn't clear when and where you're 'updating' DM w/ new swagger specs.  After i  defined the CRDs, GKE's swagger definitions get updated but DM doesn't know about them yet so i needed to run a

```
gcloud beta deployment-manager type-providers update ${CLUSTER_NAME}-provider \
    --descriptor-url=https://$MASTER_IP/openapi/v2 --api-options-file=api-options.yaml
```

to "refresh" its view.  

From the offline thread, the parsing error looks to be addressed so those games wtih updating the swagger2.json isn't necessary anymore.   i'll move forward with this PR/sample later this week w/ some more robust samples.

i'td be great if DM could 'auto update' the swagger specs somehow...(maybe a flag in the template that says 'after you run this (eg, defining the CRD itself), refresh the api reference' .

Capstan commented:
@salrashid123 I am not 'updating' DM w/ new swagger specs. The single deployment contains all the resources: cluster, type provider, crd, and crd instance. The `dependsOn` relationship orders them so that each is available by the time it's necessary. At the time the crontab resource is specified, DM fetches the OpenAPI v2 spec again, and receives the type information to actuate it.

salrashid123 commented:
@Capstan  you're right...there's no need to reload the api specs in DM (i could swear i had to do that before but it seems DM does this on its own).

Anyway, i've reverified all the scenarios here as well as the CRD setup.  I've added in two new properties as well for the cluster creation:
```yaml
    initialClusterVersion: latest
    enableKubernetesAlpha: false
```
and added a note about CRD requiring `1.14.7+` and the alphaflag to true as overrides.

---
While this does work, defining a crd and the gke cluster actuation takes a long time but thats a different issue.

@ocsig - can you start the review?

Capstan commented on examples/v2/gke/README.md:
```diff
@@ -146,30 +146,62 @@ If you manage certain resource types (`ClusterRole`, etc), you will need to gran
+If you want to deploy other dynamic k8s artifacts such as [CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions), define the collection first for that resource and then the object itself.
 
-For example, to _define_  `CustomResourceDefinitions`,
+For example, for `CustomResourceDefinitions`, create a cluster that supports [CustomResourcePublishOpenAPI](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2).  At the time of writing `11/6/19`, you need a GKE cluster version of atleast `1.14.7` plus Alpha feature gate enabled:
```
We know already that 1.15, whenever it comes out, will have CRDs available in beta, so no alpha feature gate required. Thus I think we should amend this to be something like "plus Alpha feature gate enabled if you are below 1.15".

salrashid123 replied:
done

Capstan commented on examples/v2/gke/crd/api-options.yaml:
```diff
@@ -0,0 +1,16 @@
+options:
+  validationOptions:
+    schemaValidation: IGNORE_WITH_WARNINGS
```
I tried this with FAIL as well, and was able to successfully push my demo. YMMV.

salrashid123 replied:
added `Fail` to both and worked fine still

Capstan commented on examples/v2/gke/crd/api-options.yaml:
```diff
@@ -0,0 +1,16 @@
+options:
+  validationOptions:
+    schemaValidation: IGNORE_WITH_WARNINGS
+  inputMappings:
```
We have additional mappings that are less relevant for this simple example, but more relevant when you wish to update resources. 

```
      - fieldName: metadata.resourceVersion
        location: BODY
        methodMatch: ^(put|patch)$
        value: $.resource.self.metadata.resourceVersion
```

This ensures that updates apply to the expected version, so if something has edited the resource out from under DM, you'll get an error.

```
    collectionOverrides:
    - collection: /api/v1/namespaces/{namespace}/services/{name}
      options:
        inputMappings:
        - fieldName: spec.clusterIP
          location: BODY
          methodMatch: ^(put|patch)$
          value: $.resource.self.spec.clusterIP
```

This allows you to make updates to cluster resources; otherwise you'll get an error that the clusterIP cannot be changed.

salrashid123 replied:
i relaized i wasn't even using the define `api-options.yaml` in the flow. 

i moved the api-options i had as well as the suggesion in your comment directly into the cluster creation yaml and python files

salrashid123 commented:
@thackem  if you want to review these

vmirzoyan commented:
@salrashid123 Thanks for the updated samples with support for openapi2 spec - these have been extremely helpful with migrating from the swaggerapi spec.

Although, I'm seeing abysmal delays in DM when using the openapi2 spec. Deploying ~10 services is taking over 40 minutes to complete. Anyone else experienced this?

Capstan commented:
Yes, it's a known issue related to the size of the OpenAPI v2 doc that k8s vends and the lack of caching in Deployment Manager, so it gets repeatedly downloaded, parsed, validated and then applied.

There's not really a great workaround. I'd already created custom OpenAPI v2 for GKE's `BackendConfig` and a GCF to serve it with the right host IP for the created cluster, and one could imagine adapting that to serve a cached config for just the types you need to reduce the overhead, but I've not had the time to dedicate to get that working.

Capstan commented:
/cc @andrei-scripniciuc @thackem for visibility on @vmirzoyan 's perf issues.

vmirzoyan commented:
Thanks for the tip. Your proposed workaround worked for me. I manually created a slimed down version of the spec, put it in a cloud storage bucket, created a cloud function to fetch it and changed the deployment manager type provider's descriptor URL to the cloud function endpoint.

This yields much better performance, but also creates additional complexity and overhead that I would like to avoid in the long run. Would be great if deployment manager handled this internally.
